src/wallet/rollover.rs

```diff
@@ -0,0 +1,453 @@
+            B_blinding: constants::PG.B_blinding,
+        };
+        let Com_w = pc_gens.commit(w, w_blinding);
+
+        // Step 1.4
+        let r_Q = Scalar::random(&mut rng);
+        let C_Q = tag.Q + r_Q * B;
+
+        // Step 1.5
+        let V = w_blinding * old_parameters.X_1 - r_Q * B;
+
+        // Step 1.6
+        let n_prime: Scalar = Scalar::random(&mut rng);
+
+        // Step 1.7
+        let d = Scalar::random(&mut rng);
+        let D: RistrettoPoint = d * B;
+
+        // Step 1.8
+        let r_n = Scalar::random(&mut rng);
+        let r_w = Scalar::random(&mut rng);
+        let Enc_w_B = (r_w * B, (w + r_w * d) * B);
+        let Enc_n_prime_B = (r_n * B, (n_prime + r_n * d) * B);
+
+        // Step 1.9
+        use proofs::client::*;
+
+        let (proof, points) = prove_compact(
+            &mut transcript,
+            ProveAssignments {
+                d: &d,
+                w: &w,
+                w_blinding: &w_blinding,
+                n_prime: &n_prime,
+                r_Q: &r_Q,
+                r_w: &r_w,
+                r_n: &r_n,
+                D: &D,
+                Enc_w_B_0: &Enc_w_B.0,
+                Enc_w_B_1: &Enc_w_B.1,
+                Enc_n_prime_B_0: &Enc_n_prime_B.0,
+                Enc_n_prime_B_1: &Enc_n_prime_B.1,
+                P: &tag.P,
+                V: &V,
+                Com_w: &Com_w,
+                B: &B,
+                B_blinding: &constants::B_BLINDING,
+                X_1: &old_parameters.X_1,
+                minus_r_Q: &(-r_Q),
+            },
+        );
+
+        // Step 1.10
+        Ok((
+            AwaitingResponse {
+                old_parameters: old_parameters.clone(),
+                new_parameters: new_parameters.clone(),
+                transcript,
+                w: self.w,
+                n_prime,
+                d,
+                D,
+                Enc_w_B: (points.Enc_w_B_0, points.Enc_w_B_1),
+                Enc_n_prime_B: (points.Enc_n_prime_B_0, points.Enc_n_prime_B_1),
+            },
+            Request {
+                epoch: old_parameters.epoch,
+                new_epoch: new_parameters.epoch,
+                n: self.n,
+                D: points.D,
+                Enc_n_prime_B: (points.Enc_n_prime_B_0, points.Enc_n_prime_B_1),
+                Enc_w_B: (points.Enc_w_B_0, points.Enc_w_B_1),
+                Com_w: points.Com_w,
+                P: points.P,
+                C_Q: C_Q.compress(),
+                proof,
+            },
+        ))
+    }
+}
+
+/// A response to a wallet rollover request.
+#[allow(non_snake_case)]
+pub struct Response {
+    P: CompressedRistretto,
+    Enc_Q: (CompressedRistretto, CompressedRistretto),
+    T_1: CompressedRistretto,
+    T_2: CompressedRistretto,
+    proof: proofs::issuer::CompactProof,
+}
+
+impl Request {
+    #[allow(non_snake_case)]
+    pub fn rollover<R: RngCore + CryptoRng>(
+        &self,
+        old_secret: Secrets,
+        new_secret: Secrets,
```

Comment:
```suggestion
        new_secret: &Secrets,
```
Here and above, the function should take a borrowed `Secrets` rather than taking ownership, since the `Secrets` probably shouldn't be `Copy` (it currently is, which is an unrelated bug).